python_modules/dagster/dagster/_core/definitions/sensor_definition.py

```diff
@@ -154,6 +159,120 @@ def repository_name(self) -> Optional[str]:
+
+        @sensor
+        def the_sensor(context: SensorEvaluationContext):
+            ...
+
+    """
+
+    def __init__(
+        self,
+        instance_ref: Optional[InstanceRef],
+        last_completion_time: Optional[float],
+        last_run_key: Optional[str],
+        cursor: Optional[Mapping[AssetKey, str]],
+        repository_name: Optional[str],
+        asset_keys: Sequence[AssetKey],
+        instance: Optional[DagsterInstance] = None,
+    ):
+        self._asset_keys = asset_keys
+        self.cursor_has_been_updated = False
+
+        super(MultiAssetSensorEvaluationContext, self).__init__(None, None, None, None, None, None)
+
+        self._exit_stack = ExitStack()
+        self._instance_ref = check.opt_inst_param(instance_ref, "instance_ref", InstanceRef)
+        self._last_completion_time = check.opt_float_param(
+            last_completion_time, "last_completion_time"
+        )
+        self._last_run_key = check.opt_str_param(last_run_key, "last_run_key")
+        self._cursor = check.opt_dict_param(cursor, "cursor", key_type=AssetKey, value_type=str)
+        self._repository_name = check.opt_str_param(repository_name, "repository_name")
+        self._instance = check.opt_inst_param(instance, "instance", DagsterInstance)
+
+    def latest_materializations_by_key(self, asset_keys: Optional[Sequence[AssetKey]] = None):
+        from dagster._core.events import DagsterEventType
+        from dagster._core.storage.event_log.base import EventRecordsFilter
+
+        if asset_keys is None:
+            asset_keys = self._asset_keys
+
+        cursor_dict = json.loads(self.cursor) if self.cursor else {}
+        asset_event_records = {}
+        for a in asset_keys:
+            event_records = self.instance.get_event_records(
+                EventRecordsFilter(
+                    event_type=DagsterEventType.ASSET_MATERIALIZATION,
+                    asset_key=a,
+                    after_cursor=cursor_dict.get(a.to_user_string()),
+                ),
+                ascending=False,
+                limit=1,
+            )
+
+            if event_records:
+                asset_event_records[a] = event_records[0]
+            else:
+                asset_event_records[a] = None
+
+        return asset_event_records
+
+    def materializations_for_key(self, asset_key: AssetKey, limit: int):
+        from dagster._core.events import DagsterEventType
+        from dagster._core.storage.event_log.base import EventRecordsFilter
+
+        cursor_dict = json.loads(self.cursor) if self.cursor else {}
+
+        return self.instance.get_event_records(
+            EventRecordsFilter(
+                event_type=DagsterEventType.ASSET_MATERIALIZATION,
+                asset_key=asset_key,
+                after_cursor=cursor_dict.get(asset_key.to_user_string()),
+            ),
+            ascending=True,
+            limit=limit,
+        )
+
+    def advance_cursor(
+        self, materializations_by_key
```

Comment:
When I try to add type annotations here, it should be `Mapping[AssetKey, EventLogRecord]` but that causes an import error because importing `EventLogRecord` creates a circular import. What are the options for getting around that?

Reply:
oh cool, i didn't realize those two things went together, was trying to figure out why the string type annotations didn't work since i see them everywhere. thanks!